examples/render_example/source/player.cpp

```diff
@@ -0,0 +1,314 @@
+                }
+            }
+        }
+    }
+
+    //
+    // Apply IK.
+    //
+
+    SharedPtr<Geometry> geometry = m_geometry_primitive->get_geometry();
+    if (geometry && geometry->is_loaded() && geometry->get_skeleton()) {
+        const Skeleton* skeleton = geometry->get_skeleton();
+
+        uint32_t left_thigh_index = skeleton->get_joint_index(m_left_thigh_joint);
+        uint32_t left_calf_index = skeleton->get_joint_index(m_left_calf_joint);
+        uint32_t left_foot_index = skeleton->get_joint_index(m_left_foot_joint);
+
+        uint32_t right_thigh_index = skeleton->get_joint_index(m_right_thigh_joint);
+        uint32_t right_calf_index = skeleton->get_joint_index(m_right_calf_joint);
+        uint32_t right_foot_index = skeleton->get_joint_index(m_right_foot_joint);
+
+        if (left_thigh_index != UINT32_MAX && left_calf_index != UINT32_MAX && left_foot_index != UINT32_MAX &&
+            right_thigh_index != UINT32_MAX && right_calf_index != UINT32_MAX && right_foot_index != UINT32_MAX)
+        {
+            if (is_grounded) {
+                const Vector<float4x4>& model_space_matrices = m_geometry_primitive->get_model_space_joint_pre_ik_matrices();
+
+                const float4x4& left_foot_model = model_space_matrices[left_foot_index];
+                const float4x4& right_foot_model = model_space_matrices[right_foot_index];
+
+                float4x4 global_transform_matrix(m_geometry_primitive->get_global_transform());
+
+                float4x4 left_foot_global = left_foot_model * global_transform_matrix;
+                float4x4 right_foot_global = right_foot_model * global_transform_matrix;
+
+                float3 left_foot = point_transform(float3(), left_foot_global);
+                float3 right_foot = point_transform(float3(), right_foot_global);
+
+                const float3& global_translation = get_global_translation();
+
+                std::optional<QueryResult> left_result = m_scene.raycast(float3(left_foot.x, global_translation.y + 1.f, left_foot.z), float3(0.f, -1.f, 0.f), 2.f, 1);
+                std::optional<QueryResult> right_result = m_scene.raycast(float3(right_foot.x, global_translation.y + 1.f, right_foot.z), float3(0.f, -1.f, 0.f), 2.f, 1);
+
+                float geometry_primitive_offset = 0.f;
+                
+                if (left_result) {
+                    geometry_primitive_offset = std::min(geometry_primitive_offset, left_result->position.y - global_translation.y);
+                }
+
+                if (right_result) {
+                    geometry_primitive_offset = std::min(geometry_primitive_offset, right_result->position.y - global_translation.y);
+                }
+
+                m_geometry_primitive_offset = lerp(m_geometry_primitive_offset, geometry_primitive_offset, 0.1f);
+
+                // TODO: Use linear rather than exponetial to work better on different framerate.
```

Comment:
Actually, you can easily modify exponential to be framerate-independent
http://www.viniciusgraciano.com/blog/exponential-smoothing/

Reply:
Yep, I'm aware of that. The comment is a bit misleading.